### ISSUE TITLE
HD Audio Passthrough on Nvidia Shield

### ISSUE THREAD

**Issue by Robert-D** [jellyfin/jellyfin-androidtv#281]
Hello and sorry if this is already being addressed, but it appears HD audio streams such as DTS-X and TrueHD encoded with Atmos metadata are not being passed properly to receivers/sound bars from the Nvidia Shield. It appears most people are using Kodi as a work around:

[https://www.reddit.com/r/jellyfin/comments/ewf73h/audio_passthrough_direct_on_shield_tv/](https://www.reddit.com/r/jellyfin/comments/ewf73h/audio_passthrough_direct_on_shield_tv/)

If there is anything I can do to help please let me know and sorry again if this is a duplicate.

**Comment by StereotypicalCat** [comment]
Even after #524, i am still unable to get either Atmos or DTS:X to passthrough.

**Comment by Robert-D** [comment]
Thanks for the reminder stale bot, but I would still love to see proper audio passthrough with the native player instead of using VLC.

**Comment by metronidazole** [comment]
Is it just tracks with atmos or also lossless audio? I am able to passthrough TrueHD and DTS-HD MA without issue when building this from source.

**Comment by roblav96** [comment]
More info on this wide-spread problem can be found here:
https://github.com/moneytoo/Player/issues/139

**Comment by bokysan** [comment]
Anybody tried this lately? As of late, I don't have a lot of issues using ExoPlayer, but would need to recheck if this is still an issue with the latest Jellyfin version.

**Comment by ElegyD** [comment]
I've got Jellyfin 10.7.7 with 0.12.3 on my Nvidia Shield. I'm using Exoplayer for Dolby Vision compatibility.
Dolby TrueHD passthrough works, but DTS-HD MA doesn't. Regardless of the "Bitstream DTS audio" toggled on or off.

**Comment by morkyy** [comment]
Same issue here on the shield. Most codecs passthrough just fine (including Dolby TrueHD and Atmos). But DTS seems to be very buggy. A lot of files with DTS:X/Neural X audio play fine but multiple files with DTS-HD MA and sometimes DTS:X MA do not work. Video will play fine but there will be no audio regardless of the "bitstream DTS audio" setting. Since v0.13 some files might not play at all even though they used to play with v0.12 with no sound. 

Receiver detects the correct audio format.

**Comment by roblav96** [comment]
@nielsvanvelzen Here is a diff of adding DTS-HD to ExoPlayer's `MatroskaExtractor.java`:
https://github.com/moneytoo/ExoPlayer/commit/50d7b68756e281342c85ee277ff3929a1fdf6c44

**Comment by morkyy** [comment]
To add to this, I realised that for some DTS-HD MA files where sound doesn't work there's a small crackling sound that comes up randomly. I was also playing with the settings a bit and noticed that the app would attempt to open these files with lib vlc instead of exoplayer. exoplayer seems to be playing the files correctly most of the times without transcoding. However even with selecting exoplayer there's still some files that do not play at all.

**Comment by jaytea33** [comment]
This issue has been super annoying for months. The only ones that don't work for me have EAC3 audio and everything else works fine. Exoplayer plays the video, libvlc crashes, but external VLC works perfectly

**Comment by jameskimmel** [comment]
> To add to this, I realised that for some DTS-HD MA files where sound doesn't work there's a small crackling sound that comes up randomly. 


I also experience this issue with Jellyfin Desktop on macOS. Maybe this is a deeper rooted problem. 
Let me know if you need further information to hunt down the issue.

**Comment by TraderHowell** [comment]
I am having the exact same issue on my 2019 Nvidia Shield TV Pro. I'm using Kodi as a workaround for now, but both ExoPlayer and LibVLC fail to properly passthrough DTS-HD MA when using the Jellyfin client.

**Comment by ThisIsTheOnlyUsernameAvailable** [comment]
Replying to keep this issue open.  It's a show-stopper for a lot of home theatre setups.

**Comment by WSt89** [comment]
I am facing the same issue. ExoPlayer runs very smoothly on my shield but it can't pass through DTS:MA and DTS:X. As they are very popular formats, it's support would make jellyfin a great choice for home theaters. Switching to an external player is always a work around. Can somebody explain why it's currently not possible and is somebody currently working on it?

**Comment by TraderHowell** [comment]
Also facing the same issue with Jellyfin on my Nvidia Shield. I would love if anyone can provide any guidance or insight into starting to identify a solution. Neither ExoPlayer or libVLC properly passthrough DTS-HD MA, but Dolby TrueHD passes through just fine. The amplifiers I have tested with all show DTS-HD MA for a second and then switch to lossy DTS. I have tested with so many different hardware configurations and sample formats, but nothing seems to work.

**Comment by WSt89** [comment]
My current workaround is to use Kodis Jellyfin plugin, however I would prefer to use Jellyfin app instead.

**Comment by Starfoxfs** [comment]
Same here with Nvidia Shield TV Pro 2019, i don´t wan´t to use Kodi with JellyFin Plugin.

Why you need Jellyfin then ? u can use Kodi alone to Stream Movies in DTS-HD MA
Plex does this too in Free Version.

But i like the UserInterface of JellyFin, so it would be nice if we get this HD Audio Feature for the Jellyfin Android Client.

**Comment by WSt89** [comment]
I still use Jellyfin as I have multiple clients and I want to have my libary managed on the server. I only use Kodi + Jellyfin Plugin in my home cinema.

**Comment by mikeporterinmd** [comment]
I notice that the audio codec passed from JellyFin itself is dts even for dts-hd ma.  I have figured out how to build the source and get the debugging app running on my shield 2019 pro.  Once I figure out how to debug it and what the control flow is perhaps I can make some headway.  dts-hd ma is by far the most common sound track type of quality that I have.  A real shame we can't get it to work with a nice player like this one.  We'll see.  I'm a long time C, Perl and now Python programmer.  It's going to take me a little time to come to speed in this Android environment and some of these languages and templates that are used.  I randomly tried changing some stuff, but to no effect.  One interesting thing: the new media display shows all the tracks available in a particular container.  DTS-HD MA is properly identified.  Multiple streams of magic at play here?

Mike

**Comment by Nicolas-Ger** [comment]
I also have an Nvidia Shield TV Pro 2019 and DTS HD or DTS X 7.1 is output as DTS 5.1 with the Exoplayer. With libVLC it works but with a delay that is so high that I can't set it up correctly. I used Plex before and it worked without a problem and it also plays correctly in Kodi with the Exoplayer, so it must be something to do with Jellyfin or the interaction with ffmpeg. I think everyone with the Nvidia Shield TV pro has this problem. I hope that the update to 10.9 will bring an improvement.

**Comment by mikeporterinmd** [comment]
> I also have an Nvidia Shield TV Pro 2019 and DTS HD or DTS X 7.1 is output as DTS 5.1 with the Exoplayer. With libVLC it works but with a delay that is so high that I can't set it up correctly. I used Plex before and it worked without a problem and it also plays correctly in Kodi with the Exoplayer, so it must be something to do with Jellyfin or the interaction with ffmpeg. I think everyone with the Nvidia Shield TV pro has this problem. I hope that the update to 10.9 will bring an improvement.

Just (Video) Player (On the app store) works well as an external player for JellyFin and handles DTS-HD MA 7.1 etc.  It's a nice player and very lightweight.  No where near the baggage that Kodi brings in.  I think there are some things like resume that might not work correctly... not sure.  I don't know why the JellyFin Android developers don't work with the Just (Video) Player author to merge the code bases.  I took a look at the code, but there is no real documentation in the JellyFin app, so lacking experience in Android development, merging the code wasn't worth it for me.  Not when I can just use the other player.

**Comment by ThisIsTheOnlyUsernameAvailable** [comment]
Bug still exists.  Not stale.

**Comment by armageddon421** [comment]
In Case this helps narrowing down the Problem:
I Just found that disabling the framerate matching in the jellyfin settings solves a lot of Problems for me, where previously the AVR would only receive Multi Channel audio instead of Atmos or DTS.

**Comment by buhtig20** [comment]
Bug still exists for me with DTS-X even after disabling refresh rate switching in Jellyfin settings! However, Dolby Atmos is passing through.

**Comment by OlJohnny** [comment]
For the audio transcoding issue stemming from the "Refresh Rate Swithcing" feature, see issue #4067.
---------------
My issues are completely fixed by turning that feature off, even with DTS-X and an up to date Nvidia Shield TV 2019 (the tube) directly connected an HDMI-2.1 Port of an Denon AVR X1700H.

@buhtig20 at first I had the similar issue, that DTS codecs weren't getting passed through.
My audio chain had looked like this: Nvidia Shield TV -> LG OLED C1 -> Denon ARV X1700H
Turns out that the **LG OLED doesn't support DTS codecs**, even when only passing through. Connecting the Shield directly to the Denon fixed that issue.

**Comment by buhtig20** [comment]
> ## For the audio transcoding issue stemming from the "Refresh Rate Swithcing" feature, see issue [#4067](https://github.com/jellyfin/jellyfin-androidtv/issues/4067).
> My issues are completely fixed by turning that feature off, even with DTS-X and an up to date Nvidia Shield TV 2019 (the tube) directly connected an HDMI-2.1 Port of an Denon AVR X1700H.
> 
> [@buhtig20](https://github.com/buhtig20) at first I had the similar issue, that DTS codecs weren't getting passed through. My audio chain had looked like this: Nvidia Shield TV -> LG OLED C1 -> Denon ARV X1700H Turns out that the **LG OLED doesn't support DTS codecs**, even when only passing through. Connecting the Shield directly to the Denon fixed that issue.

Thank you but I don't believe it's a setup issue. The Emby and Plex apps on the Shield passthrough DTS-X when playing the same file.

**Comment by badaz** [comment]
I'm quite interested in getting a Ugoos AM8 pro for use with slimbox tv android 11 tv firmware (I'm not a big Kodi fan so I'd prefer not having to use Coreelec), but I'm a bit afraid of what I'm reading here... Is this a generic jellyfin + android tv 11 issue or does it only affect the Shield?

**Comment by OlJohnny** [comment]
> Is this a generic jellyfin + android tv 11 issue or does it only affect the Shield?

Seems to be more evident on the Shield.
However a fix for this issue seems to be on the way (for now still waiting on a new release upstream: https://github.com/androidx/media/issues/2258).

**Comment by WSt89** [comment]
The fix you are refering to does not seem to solve the issue described here, as it's fixing an unwanted PCM stream. However, in this thread we cannot pass through DTS-HD MA and DTS:X, because it's being streamed as DTS 5.1 only (Dolby TrueHD and Atmos works though).

**Comment by ThisIsTheOnlyUsernameAvailable** [comment]
Not stale.

**Comment by Starfoxfs** [comment]
Same problem here on a Nvidia Shiel TV Pro 2019, no Passthrough DTS-HD MA, Dolby True HD. Jellyfin native does not work to passthrough this Audio.

**Comment by RyuzakiKK** [comment]
Support for DTS-HD MA and DTS:X has been added to media3 (exoplayer), and should be included when they'll tag a new release (after the current 1.8.0)

https://github.com/jellyfin/jellyfin-androidtv/issues/3166#issuecomment-3163825722

> Same problem here on a Nvidia Shiel TV Pro 2019, no Passthrough DTS-HD MA, Dolby True HD

I have the same Nvidia Shield Pro 2019 and Dolby True HD works fine. I'm not sure what's the issue you are facing here.

**Comment by grasshide** [comment]
I mostly experience the problem that Atmos is played as Multi Channel when I play content with high resolution (4k remux for example). Interestingly emby does not have that problem.

**Comment by Blend831** [comment]
I have the same issue described here - NVIDIA Shield TV Pro (2019) > Denon AVR + LG TV. Attempting to play Atmos and DTS content triggers transcoding at the server, does not passthrough.

I've found a workaround for now which is that Wholphin has an option to disable FFmpeg decoding at the player, which then enables proper directplay+passthrough of uncompressed audio via ExoPlayer: https://github.com/damontecres/Wholphin/issues/732

**Comment by grasshide** [comment]
I tried to collect some information that is reproducible and that others might be able to confirm.
All tests are done with Jellyfin 10.11.6 and Jallyfin Android TV 0.19.7
I use the Shield (non Pro) and an Epson UHD projector, both connected to a Denon X3800H.
Bitstream Dolby Digital audio is enabled, Refresh rate switching is Scale on TV, Audio output set to direct, Video start delay: 5s (could be more since the projector takes 10-15s to adjust to framerate, resolution, HDR, etc.).

I took the samples from here: https://kodi.wiki/view/Samples

Dolby ATMOS 'Amaze' Demo (M2TS @ 1080p/24)
- No 8 under "HD/object-based Audio Test Clips"
With or without option "Allow audio playback that requires transcoding" enabled:
- Sound is present (AVR says Signal: Dolby Atmos - TrueHD)
- With Emby, sound is present (AVR says Signal: Dolby Atmos - TrueHD)

HDR10+ Profile B HEVC 10-bit 23.976 Sample (MKV with TrueHD ATMOS audio)
- No 12 under "4K (UltraHD) & HDR Formats" (scene in police station)
- Option "Allow audio playback that requires transcoding" disabled:
  - No Sound (AVR says Signal PCM; Dolby Surround with Stereo)
- Option "Allow audio playback that requires transcoding" enabled:
  - Sound is present (AVR says Signal PCM; Multi Channel In with 5.1 channels)
- With Emby, sound is present (AVR says Signal: Dolby Atmos - TrueHD)

Hybrid HDR10/Dolby Vision Sample (MKV with TrueHD ATMOS audio)
- No 14 under "HD/object-based Audio Test Clips" (scene in Sicily)
- Option "Allow audio playback that requires transcoding" disabled:
  - No Sound (AVR says Signal PCM; Dolby Surround with Stereo)
- Option "Allow audio playback that requires transcoding" enabled:
  - Sound is present (AVR says Signal PCM: Multi Channel In with 5.1 channels)
- With Emby, sound is present (AVR says Signal: Dolby Atmos - TrueHD)

**Comment by d4v3d** [comment]
Hi all, digging up an older thread here but I was able to get DTS-HD MA audio (specifically the LOTR EE blu-ray rips) working on my Shield Pro by going into Developer Options and disabling 'Prefer FFmpeg for audio playback'. This is on the latest Android client - version 0.19.8.

I'm testing other media as well and so far haven't noticed any issues.

Hope this helps someone!
Dave

**Comment by mikeporterinmd** [comment]
I am not the original submitter, but I can confirm the issue still exists.

**Comment by patricks1** [comment]
@mikeporterinmd What version are you on? It's solved for me, as mentioned in #3166. Maybe it's working for me because I pulled the master branch and installed that on my Shield. Are you willing/able to do that?

**Comment by misomarcell** [comment]
This is still an issue in 0.19.10 on a 2019 nvidia shield. TrueHD is being converted to PCM, while the same file plays perfectly fine on Kodi. Same with most Dolby Vision profiles.

**Comment by mikeporterinmd** [comment]
No, I am not willing.  This comment stream has been active for at least a year.  I learned how to side load, tried testing and discovered a lot of issues.  Willing to help, but comments in code help just-a-little.  A number of external players have the fix, too.

**Comment by patricks1** [comment]
@misomarcell @mikeporterinmd I mean... I think the purpose of GitHub issues is so the devs can keep track of what they need to fix in the code, and the code is fixed. @nielsvanvelzen said [the v0.20 release will include the relevant commits](https://github.com/jellyfin/jellyfin-androidtv/issues/3166#issuecomment-3668965417). You could try asking them what the timeline looks like for tagging that release, but there's nothing left to fix in the code. (I'm just trying to help, not arguing.)

At the risk of sidetracking this discussion, what issues did you have with the master branch @mikeporterinmd? I pulled it up to [`913beda91`](https://github.com/jellyfin/jellyfin-androidtv/tree/913beda916d311617a0f3bb43a50a57c4ec968f9) and didn't notice any issues. You could open a new issue for any problem you have, wait for that to get fixed, and then pull master after that point. Or you could pull master, fix the issues on your own, and then send your apk to your Shield. I'm just trying to give you options.
